Prevent HSpec from messing with ImpSpec colors

lehins opened the pull request:
# Description

@neilmayhew This is the piece you are missing for the first point in your PR: #4497 

> hspec forces a red colour on every line when it outputs the text of a failure message.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

neilmayhew commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -817,7 +818,19 @@ instance (ShelleyEraImp era, Arbitrary a, Show a, Testable prop) => Example (a -
+      resultStatus -> resultStatus
+    adjustFailureReason = \case
+      H.Reason reason -> H.ColorizedReason reason
+      failureReason -> failureReason
```
I find this a bit more readable, but you might not:

```suggestion
      H.Failure mLoc (H.Reason reason) -> H.Failure mLoc (H.ColorizedReason reason)
      resultStatus -> resultStatus
```

lehins replied:
No longer applicable

neilmayhew reviewed (approved):
This is great! Well done for discovering this.

I thought perhaps there might still be a problem with the `ExpectedButGot` constructor, but I see that HSpec is doing its own diff in that case, and wouldn't be using a tree-diff of ours.

neilmayhew commented:
We also need to reverse the addition of `--no-color` in GHA. However, I can take that out of my PR since it's not merged yet and that change is a separate commit.

lehins committed:
Handle color output with HSpec

lehins commented:
@neilmayhew I've changed the original fixup that you've reviewed, because it was only applicable to ImpSpec, while all other usages of color still had problems. 

I did get to the bottom of it and it turned out if we report any failure that depends on color output with `ColorizedReason`, then the color will be preserved and not botched by hspec.

Failure reporting for ImpSpec was quite wrong too, which is also fixed in this PR.

neilmayhew reviewed (approved):
This turned out to be surprisingly complex! Looks very solid, though. Nice to have it taken care of properly.

I've taken out the CI changes in my PR. I've also rebased it, which is why this branch now has conflicts. LMK if you'd like me to fix this for you.

aniketd commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -1120,12 +1124,11 @@ logAcceptedRatio aId = do
-      , "----- ACCEPTED RATIOS -----"
-      , "DRep accepted ratio:\t\t" <> show dRepRatio
-      , "Committee accepted ratio:\t" <> show committeeRatio
-      , "SPO accepted ratio:\t\t" <> show spoRatio
```
Very cool! 🚀
